File: backend/src/utils/age.ts
/**
 * Majorité — **source unique**, et seule autorité côté serveur.
 *
 * L'âge se calculait jusqu'ici dans un seul contrôleur, par division :
 *
 *     Math.floor((Date.now() - dob.getTime()) / (365.25 * 86400000))
 *
 * Approximation qui se trompe d'un jour autour de l'anniversaire, selon les
 * années bissextiles traversées. Sur un seuil légal, se tromper d'un jour n'est
 * pas anodin : c'est exactement le jour où quelqu'un devient majeur.
 *
 * Surtout, ce calcul ne s'appliquait qu'à `PATCH /profile`, et seulement
 * lorsqu'une date était transmise (`if (birth_date)`). Un compte dont la date
 * arrivait par un autre chemin n'était jamais contrôlé, et le reste de l'API
 * se contentait de vérifier qu'une date **existait**.
 */
export const AGE_MINIMUM = 18;

/** Âge révolu, en années pleines, par arithmétique calendaire. */
export function ageRevolu(naissance: Date, maintenant = new Date()): number {
  let age = maintenant.getUTCFullYear() - naissance.getUTCFullYear();

  const moisEcoule = maintenant.getUTCMonth() - naissance.getUTCMonth();
  const jourEcoule = maintenant.getUTCDate()  - naissance.getUTCDate();

  // L'anniversaire de cette année n'est pas encore passé.
  if (moisEcoule < 0 || (moisEcoule === 0 && jourEcoule < 0)) age -= 1;

  return age;
}

/**
 * Date valide **et** majeur.
 *
 * Une date invalide, absente ou située dans le futur ne peut pas établir la
 * majorité : dans le doute, on refuse.
 */
export function estMajeur(naissance: Date | null | undefined,
                          maintenant = new Date()): boolean {
  if (!naissance) return false;
  const d = naissance instanceof Date ? naissance : new Date(naissance);
  if (isNaN(d.getTime())) return false;
  if (d.getTime() > maintenant.getTime()) return false;
  return ageRevolu(d, maintenant) >= AGE_MINIMUM;
}

File: mobile_new/lib/shared/utils/age.dart
/// Âge révolu, par arithmétique calendaire.
///
/// L'écran du compte calculait l'âge en divisant un nombre de jours par
/// 365,25 :
///
///     ((DateTime.now().difference(d).inDays) / 365.25).floor()
///
/// L'approximation se trompe d'un an **le jour de l'anniversaire**, selon les
/// années bissextiles traversées : sur trente dates éprouvées, vingt donnaient
/// un an de moins le jour même. Quelqu'un qui a vingt et un ans lisait
/// « 20 ans », le seul jour de l'année où il regarde cette ligne.
///
/// ── Ce calcul existait déjà, corrigé, ailleurs ─────────────────────────────
///
/// `backend/src/utils/age.ts` a remplacé cette même formule par ce même calcul,
/// avec son banc, parce qu'elle décidait de la majorité : se tromper d'un jour
/// sur un seuil légal, c'est se tromper exactement le jour où quelqu'un devient
/// majeur. La règle était donc écrite, corrigée et éprouvée — d'un seul côté.
///
/// Le serveur reste la seule autorité sur la majorité ; l'application ne fait
/// qu'afficher. Mais afficher faux reste afficher faux.
int ageRevolu(DateTime naissance, [DateTime? maintenant]) {
  final m = maintenant ?? DateTime.now();

  var age = m.year - naissance.year;

  final moisEcoule = m.month - naissance.month;
  final jourEcoule = m.day - naissance.day;

  // L'anniversaire de cette année n'est pas encore passé.
  if (moisEcoule < 0 || (moisEcoule == 0 && jourEcoule < 0)) age -= 1;

  return age;
}
